src/java.base/share/classes/jdk/internal/foreign/abi/BindingSpecializer.java

```diff
@@ -77,6 +81,8 @@ public class BindingSpecializer {
         = GetPropertyAction.privilegedGetProperty("jdk.internal.foreign.abi.Specializer.DUMP_CLASSES_DIR");
     private static final boolean PERFORM_VERIFICATION
         = GetBooleanAction.privilegedGetProperty("jdk.internal.foreign.abi.Specializer.PERFORM_VERIFICATION");
+    private static final int SCOPE_DEDUP_DEPTH
+            = GetIntegerAction.privilegedGetProperty("jdk.internal.foreign.abi.Specializer.SCOPE_DEDUP_DEPTH", 2);
```

Comment:
A two-comparison logic seems like a good compromise. Note this doesn't really mean a two-element cache - it is still possible to have e.g. the first two arguments with the same scope and then a third argumemt with another scope - in which case the third argument won't be deduped. This is a trade-off - to keep the generated code simple we need to have a 1-1 mapping between by-ref parameters and locals.